localstack-core/localstack/services/lambda_/event_source_mapping/pollers/stream_poller.py

```diff
@@ -135,14 +141,30 @@ def poll_events_from_shard(self, shard_id: str, shard_iterator: str):
             record_age_in_seconds = now - arrival_timestamp_of_last_event
             if record_age_in_seconds > maximum_record_age_in_seconds:
                 abort_condition = "RecordAgeExpired"
-        matching_events = self.filter_events(polled_events)
+
+        # TODO: implement format detection behavior (e.g., for JSON body):
+        #  https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-pipes-event-filtering.html
+        #  Check whether we need poller-specific filter-preprocessing here without modifying the actual event!
+        # convert to json for filtering (HACK for fixing parity with v1 and getting regression tests passing)
+        # localstack.services.lambda_.event_source_listeners.kinesis_event_source_listener.KinesisEventSourceListener._filter_records
+        # TODO: explore better abstraction for the entire filtering, including the set_data and get_data remapping
+        #  We need better clarify which transformations happen before and after filtering -> fix missing test coverage
+        parsed_events = self.pre_filter(polled_events)
+        # TODO: advance iterator past matching events!
+        #  We need to checkpoint the sequence number for each shard and then advance the shard iterator using
+        #  GetShardIterator with a given sequence number
+        #  https://docs.aws.amazon.com/kinesis/latest/APIReference/API_GetShardIterator.html
+        #  Failing to do so kinda blocks the stream resulting in very high latency.
```

Comment:
Why would the shard iterator not advance if we received the messages? Even if none of them match, we still set the new shard iterator, right?

Reply:
That only happens in (partial) error case. For example, if out of 10 events, the first five do not match the filter but others in the batch cause a (partial) error. We need to figure out whether AWS forwards the iterator passing the first non-matching events because they would be filtered out in a future retry anyway.